flows.go

```diff
@@ -158,18 +145,29 @@ func FlowFromEndpoints(src, dst Endpoint) (_ Flow, err error) {
 
 // FastHash provides a quick hashing function for a flow, useful if you'd
 // like to split up flows by modulos or other load-balancing techniques.
-// It uses a variant of Fowler-Noll-Vo hashing, and is guaranteed to collide
-// with its reverse flow.  IE: the flow A->B will have the same hash as the flow
-// B->A.
+// It  is guaranteed to collide with its reverse flow.  IE: the flow A->B
+// will have the same hash as the flow B->A.
+//
+// It creates the hash by first, for each endpoint, xor-ing the bytes of the
+// address in a 32-bit round robin fashion (hash ^= addr[0],hash ^= addr[1]<<8,
+// ...,hash ^= addr[5]<<8). The result for each endpoint is then concatenated
+// together in an order independent way. For addresses less than or equal to
+// 32-bits this guarantees collision free hashes. IE: the flows A->B and B->A
+// will have a hash that no other combination of addresses can produce.
 //
 // The output of FastHash is not guaranteed to remain the same through future
 // code revisions, so should not be used to key values in persistent storage.
-func (f Flow) FastHash() (h uint64) {
-	// This combination must be commutative.  We don't use ^, since that would
-	// give the same hash for all A->A flows.
-	h = fnvHash(f.src[:f.slen]) + fnvHash(f.dst[:f.dlen])
-	h ^= uint64(f.typ)
-	h *= fnvPrime
+func (f Flow) FastHash() (a uint64) {
+	var b uint64
+	for i := 0; i < f.slen; i++ {
+		a ^= uint64(f.src[i]) << (8 * (uint(i) % 4))
```

Comment:
I'm worried about this approach for large values, specifically IPv6 addresses.  As well as being > 4 bytes, they're also somewhat likely to have some "structure" to their bits set/unset.  For example, I could see a network having blah:1:1:1:1 and blah:2:2:2:2, which would evaluate to the same hash due to the XOR canceling out the first 1:1 with the second, and doing the same for the first/second 2:2.

Could we go with your original approach of just doing the 32-bit method for values <= 4 bytes?  I think it might be easily simplified as:

```
if f.slen <= 4 && f.dlen <= 4 {
  // LittleEndian in case only the first or first 2 bits are set.  Assumes bits after xlen are all zeros.
  a := binary.LittleEndian.Uint32(a.src[:4])
  b := binary.LittleEndian.Uint32(a.dst[:4])
  if a > b { a, b = b, a }
  return uint64(a) << 32 | uint64(b)
}
```

Reply:
Completely understand, there are two main issues: 
1. Inline budget 
2. The Hash collision cases

**Inline budget**
If both FNV and Direct Hashing are implemented, it will throw the FastHashes compile cost over the inline budget and give them a 10x slowdown. I have tried every way I can think of to combine them and have not even gotten close. Adding the change you suggested results in:
```
./flows.go:175:6: cannot inline Flow.FastHash: function too complex: cost 146 exceeds budget 80
```

**The Direct Hash collision cases**
You're right with regards to Flow cases:
* `fdff::1:1:1:1` ->`fdff::2:2:2:2`
* `fdff::2:2:2:2` ->`fdff::3:3:3:3`

My only argument is that the FNV still has issues with IPv6 flows as well. They aren't as structural but they occur in common analysis. I am sure I can find a collision with addresses closer together, but just porting my initial TCP Port collision from #860 to IPv6:
* `fdff::1b74` -> `fdff::fccd`
* `fdff::1b75` -> `fdff::fcce`

It is very hard for these smaller hash functions to combat all the different ways you can deploy IPv6 and are forced to analyze it. The best way may be to just handle IPv6 hashing in the `layer` package and allow for different Flow Label functions for different IPv6 situations.